lib/puma/server.rb

```diff
@@ -476,19 +480,19 @@ def process_client(client)
 
             requests += 1
 
-            # As an optimization, try to read the next request from the
-            # socket for a short time before returning to the reactor.
-            fast_check = @status == :run
+            Thread.pass
 
-            # Always pass the client back to the reactor after a reasonable
-            # number of inline requests if there are other requests pending.
-            fast_check = false if requests >= @max_fast_inline &&
-              @thread_pool.backlog > 0
+            client.reset
 
-            next_request_ready = with_force_shutdown(client) do
-              client.reset(fast_check)
+            # This indicates that the socket has pipelined (multiple)
+            # requests on it, so process them
+            next_request_ready = if client.has_buffer
```

Comment:
Request: Change this variable to indicate it's specifically related to pipelines maybe `pipeline_request_ready`

Reply:
It's not clear to me if the term "pipeline" here is a Puma-internal concept or if it's referring to [HTTP pipelining](https://en.wikipedia.org/wiki/HTTP_pipelining). If the latter, _technically_ Puma has already prepared and sent a response to the last request. Thus, the next request being ready doesn't mean the requests have been pipelined. It just means the client already sent another request on that connection.

Apologies if this is just me misunderstanding the term.

Reply:
@slizco

Sorry, wasn't sure what to call it.  I did say pipelined **requests**, so I'm referring to multiple requests sent one after the other, as opposed to request -> response -> next request...

Puma does not concurrently generate responses to pipelined requests, so there isn't any issue with the responses being re-ordered. I added the logic so that a keep-alive connection with pipelined **requests** will 'monopolize' the thread to complete all their responses.

Reply:
Naming aside my biggest worry would be that monopolization could result in a denial of service. I don't know what natural bounds the system has or enforces. For example: If someone can send 1000 requests in a way that can be fully read and buffered by the reactor then we would be opening ourselves up to a denial of service attack. I don't know if that's possible or what the conditions are necessary to trigger this codepath though.

My reading and understanding is that in this context if the body already contained multiple requests that this is where they'll be handled, but if new requests come in on the same connection, the thread will not read from the socket here and it must go back to the reactor (after this change).

My original gut reaction was to want to get rid of this special case, however adding an early return results in test failures https://gist.github.com/schneems/2a4ab615a03db5ac845fbac1542e7122. Can you say more about the desire for the monopoly? Are there some natural limits on the system that would prevent a denial of service?

Alternatively: Could we throw this case back to the reactor somehow?